zulip/tests/test_api.py

```diff
@@ -0,0 +1,95 @@
+import json
+import unittest
+import responses
+import zulip
+import urllib
+
+from typing import Any, Tuple, Dict
+from unittest import TestCase
+
+class TerminationException(Exception):
+    pass
+
+class TestAPI(TestCase):
+
+    @responses.activate
+    def test_add_reaction(self) -> None:
+        def request_callback(request: Any) -> Tuple[int, Dict[str, str], str]:
+            params = {}
+            for param in request.body.split("&"):
+                key, value = param.split("=")
+                params[key] = urllib.parse.unquote(value)
+            assert "emoji_name" in params or "emoji_code" in params
+            return (200, {}, json.dumps({'result': 'success', 'msg': ''}))
```

Comment:
I don't understand this function -- why do we need this parsing logic?

Reply:
This function needs some work as well as I understand. This test is primarily for testing whether the payload provided is properly parsed, the URL is correct, and all the It accepts and sends a proper response, as a part of the normal API workflow.

This test makes sure all intermediate functions for an API call like `do_api_query` and `call_endpoint`, work correctly.